mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodEventHandler.m

```diff
@@ -723,98 +452,102 @@ -(NSString*)readContext:(NSEvent *)event forClient:(id) client {
-    if (operation.isForCompositeAction) {
-      [self executeCompositeOperation:operation keyDownEvent:event client:client];
+-(BOOL)applyKeyOutputToTextInputClient:(CoreKeyOutput*)output keyDownEvent:(nonnull NSEvent *)event client:(id) client {
+  BOOL handledEvent = YES;
+  
+  if (output.isInsertOnlyScenario) {
+    [self.appDelegate logDebugMessage:@"KXMInputMethodHandler applyOutputToTextInputClient, insert only scenario"];
+    [self insertAndReplaceTextForOutput:output client:client];
+  } else if (output.isDeleteOnlyScenario) {
+    if ((event.keyCode == kVK_Delete) && output.codePointsToDeleteBeforeInsert == 1) {
+      // let the delete pass through in the original event rather than sending a new delete
+      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler applyOutputToTextInputClient, delete only scenario"];
+      handledEvent = NO;
     } else {
-      [self executeSimpleOperation:operation keyDownEvent:event];
+      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler applyOutputToTextInputClient, delete only scenario"];
+      [self sendEvents:event forOutput:output];
+    }
+  } else if (output.isDeleteAndInsertScenario) {
+    if (self.apiCompliance.mustBackspaceUsingEvents) {
+      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler applyOutputToTextInputClient, delete and insert scenario with events"];
+      [self sendEvents:event forOutput:output];
+    } else {
+      [self.appDelegate logDebugMessage:@"KXMInputMethodHandler applyOutputToTextInputClient, delete and insert scenario with insert API"];
+      // directly insert text and handle backspaces by using replace
+      [self insertAndReplaceTextForOutput:output client:client];
     }
   }
-
-  return result.handledEvent;
+  
+  return handledEvent;
 }
 
-/**
- * If we need to do something in response to a single action, then do it here.
- * Most of the interesting work is done in executeCompositeOperation, and
- * much of the other work is done when updating the context.
- */
--(void)executeSimpleOperation:(KMActionOperation*)operation keyDownEvent:(nonnull NSEvent *)event {
-  CoreAction *action = operation.action;
-  if (action != nil) {
-    [self.appDelegate logDebugMessage:@"KXMInputMethodHandler executeSimpleOperation: %@", operation];
+-(void)applyNonTextualOutput:(CoreKeyOutput*)output {
+  [self persistOptions:output.optionsToPersist];
+  
+  [self applyCapsLock:output.capsLockState];
+  
+  if (output.alert) {
+    NSBeep();
   }
-  switch(action.actionType) {
-    case AlertAction:
-      NSBeep();
-      break;
-    case PersistOptionAction:
-      if(action.key && action.value) {
-        [self.appDelegate logDebugMessage:@"PersistOptionAction calling writePersistedOptions, key: %@, value: %@", action.key, action.value];
-        [self.appDelegate writePersistedOptions:action.key withValue:action.value];
+}
+
+-(void) persistOptions:(NSDictionary*)options{
+  if(options.count>0) {
```

Comment:
Do we need this test? Because the loop will be a no-op anyway

Reply:
removed it